Fix various issues found by 'ansible-lint' in DebOps roles

drybjed opened the pull request.

drybjed committed:
Fix ansible-lint error E405 in DebOps roles

drybjed committed:
Add 'ansible-lint' configuration file

drybjed committed:
Fix ansible-lint error E304 in DebOps roles

drybjed committed:
Fix ansible-lint error E206 in DebOps roles

drybjed committed:
Fix ansible-lint error E202 in DebOps roles

drybjed committed:
Fix ansible-lint error E104 in DebOps roles

drybjed committed:
Fix ansible-lint error E501 in DebOps roles

drybjed committed:
[debops.owncloud] Avoid privilege escalation

The 'unarchive' task in the 'debops.owncloud' role was executed as
an unprivileged user, however the specified 'owner' and 'group'
parameters forced the module to request privilege escalation to root.
This doesn't work because the 'www-data' UNIX account under which the
task is executed is not authorized for privilege escalation. The task
exited with the "Timeout (12s) waiting for privilege escalation prompt"
error.

To fix this, the 'unarchive' task is executed as privileged, and the
'unarchive' module will use its own mechanism to change the owner and
group of extracted files.

drybjed committed:
Fix typo in variable name

drybjed committed:
Fix ansible-lint error E601 in DebOps roles

drybjed committed:
Fix ansible-lint error E302 in DebOps roles

drybjed committed:
Fix ansible-lint error E301 in DebOps roles

drybjed committed:
Fix ansible-lint error E401 in DebOps roles

drybjed committed:
Fix ansible-lint error E602 in DebOps roles

drybjed committed:
Fix ansible-lint error E305 in DebOps roles

drybjed committed:
Fix ansible-lint error E403 in DebOps roles

drybjed committed:
Fix ansible-lint error E303 in DebOps roles

ypid commented:
Give me a few days :)

drybjed commented:
@ypid Actually, I would want to finish fixing `ansible-lint` issues (317 lines too long left) and prepare next release of DebOps soon. This is so that on Galaxy all of the issues detected currently by the import process are fixed.

I tested these changes multiple times while cleaning up the code, and I'm confident that there are no changes vs the current code. I'll wait with merging this PR while cleaning up the long lines, it should take a day or two. Of course, `ansible-lint` will be then added to the test pipeline on Travis-CI.

drybjed committed:
Fix ansible-lint error E204 in DebOps roles

drybjed committed:
[ci] Check DebOps roles using ansible-lint

drybjed commented:
@ypid All fixes for issues found by `ansible-lint` are done. I plan to merge this PR on Sunday. :-)

https://xkcd.com/1070/

ypid commented on ansible/roles/debops.ferm/tasks/main.yml:
```diff
@@ -23,7 +25,7 @@
     state: 'directory'
     owner: 'root'
     group: 'adm'
-    mode: '2750'
```
This should probably have one `0` less :)

ypid commented on ansible/roles/debops.gitusers/tasks/git-shell.yml:
```diff
@@ -37,19 +38,20 @@
     state: 'directory'
     owner: 'root'
     group: '{{ item.0.group | default(item.0.name + gitusers_name_suffix) }}'
-    mode: '2775'
```
Same. I am guessing some `git ls-files -z "$(git rev-parse --show-toplevel)" | xargs --null -I '{}' find '{}' -type f -print0 | xargs --null sed --in-place --regexp-extended 's/mode: .*.../\1/g;'` magic :)

drybjed replied:
The leading zero in the value works fine with the `file` Ansible module.

ypid commented on ansible/roles/debops.opendkim/tasks/main.yml:
```diff
@@ -149,6 +152,6 @@
     owner: '{{ opendkim__user }}'
     group: '{{ opendkim__postfix_group }}'
-    mode: '2750'
+    mode: '02750'
```
Same.

drybjed replied:
The leading zero in the value works fine with the `file` Ansible module.

ypid commented on ansible/roles/debops.rails_deploy/tasks/deploy.yml:
```diff
@@ -139,20 +150,25 @@
+  command: rm -f public/index.html
+  args:
+    chdir: '{{ rails_deploy_src }}'
+    warn: False
```
This can also be expressed as file: state=absent. Then you also don’t need the `warn: False` ;-)

drybjed replied:
I know, but to fix `ansible-lint` issues in this PR I wanted to do as few changes to the existing code as possible, since it's a large set of patches. Tasks like this can be converted later to proper Ansible modules.

ypid replied:
Then maybe add a FIXME note to the files. My reasoning is that `ansible-lint` makes you aware of things that can be improved. When you do as little changes as possible if just feels like you hack the linter in ways that it will stop complying for example by disabling warnings while there is a proper way to do it. This is a trivial fix/change.

ypid commented on ansible/roles/debops.rails_deploy/tasks/deploy_keys.yml:
```diff
@@ -14,15 +14,20 @@
+                 --data '{{ rails_deploy_key_data | to_nice_json }}'
+                 https://api.github.com/repos/{{ rails_deploy_git_account }}/{{ rails_deploy_git_repo }}/keys"
+  args:
+    warn: False
```
It should be possible to not depend on `curl` here and do it with a Ansible module.

drybjed replied:
This is the same reason as with the `rm` and the `command` module - my aim here was to fix the `ansible-lint` error with minimum changes to the existing code.

ypid replied:
Same as I explained with the `rm` and the `command` module except that I am more on your side here because this task is more complex than a simple file: state=absent. Maybe also add FIXME notes so that people working on this role the next time and test it can fix the task in one go.

ypid commented on ansible/roles/debops.rails_deploy/tasks/deploy_keys.yml:
```diff
@@ -31,7 +36,12 @@
+                 https://{{ rails_deploy_git_host }}/api/v3/projects/{{
+                   rails_deploy_register_gitlab_response.json.id }}/keys"
+  args:
+    warn: False
```
It should be possible to not depend on `curl` here and do it with a Ansible module. Also the `--insecure` gives me nightmares.

drybjed replied:
Agreed, the `rails_deploy` role is unfortunately neglected, I'm not using it personally. Perhaps some kind of survey or poll to find what roles are popular among users could be interesting to do. Food for thought.

ypid replied:
Fair enough. Maybe a FIXME note can be added for now as this really needs fixing.

drybjed replied:
The whole `debops.rails_deploy` role needs a major overhaul at some point, since lots of the other roles that it depends on changed over time. We will get to it eventually. :-)

ypid commented on ansible/roles/debops.sftpusers/tasks/sites.yml:
```diff
@@ -20,7 +20,7 @@
     state: 'directory'
     owner: 'root'
     group: '{{ item.group | default(item.name + sftpusers_name_suffix) }}'
-    mode: '2775'
```
Please double check the mode parameter in this file.

drybjed replied:
The value with leading zero works fine with the `file` Ansible module.

ypid reviewed (changes requested):
LGTM. Thanks for your work! A few comments are attached.

ypid commented on ansible/roles/debops.ferm/tasks/main.yml:
```diff
@@ -23,7 +25,7 @@
     owner: 'root'
     group: 'adm'
-    mode: '2750'
+    mode: '02750'
```
This should probably have one 0 less :)

drybjed replied:
This is a fix for the E202 error in `ansible-lint`. The `file` Ansible module seems to work as expected with this leading zero.

ypid replied:
Hm, but is the leading 0 needed. It looks wired. In all the other places we use 4 digits. My point here is consistency in the code base.

drybjed replied:
In other cases with 4 digits the attributes are usually `0755` or `0644` which is "natural". In this case and similar, the role sets SGID attribute, but the notation, at least in Ansible, still hinges on `0` being the lead digit which signifies an octal number. Remember when in some cases an unquoted attribute mask resulted in a weird permissions because the number was interpreted wrong by the `file` module? I bet that unquoted number with `0` as a prefix would be interpreted correctly as an octal. Using `0` for octal numbers for me seems more consistent than using 4 digits for attribute values, at least when interpreting values correctly by the software.

ypid replied:
I like to use 4 digits for the mode to be explicit so that it is clear that I did not forget a special permission. Interestingly the only use of 5 digits currently in master is apparently from me in the `debops.apt_cacher_ng` role (`ag 'mode:[^0-9\n]*[2-7][0-7]{3}'`).

The mode parameters with 4 digits not starting with 0 currently in master (that you fix in this PR), it has not caused issues because they are all quoted. I am still in favor of your point that you want to avoid permission issues. This only stands against that we have not seen/used this before DebOps really.

Thanks for explaining. Lets go with 5 digits in case the special file/dir permissions are used.

Using `ag 'mode:[^0-9]*[2-7][0-7]{3}'` I found one additional instance of this in `ansible/roles/debops.icinga_web/tasks/main.yml` which was not found by `ansible-lint`. Can you update this also?

drybjed replied:
@ypid Thanks, done.

drybjed commented:
@ypid OK, answered all your comments, waiting for response. :-)

ypid commented:
@drybjed Done that, sorry for the delay ;-)

drybjed committed:
[debops.rails_deploy] Replace 'command' module

The task that removes the temporary deploy page should be done via the
'file' module instead of a general-purpose 'command' module.

drybjed committed:
[debops.rails_deploy] Add FIXME notes to tasks

drybjed commented:
@ypid I added the `FIXME` notes to `debops.rails_deploy` tasks, and used the `file` module instead of simple `rm` command.

ypid commented:
@drybjed Thanks! Can you comment on the file module and mode parameter discussion?

drybjed commented:
@ypid I did: https://github.com/debops/debops/pull/600#discussion_r247350928

ypid commented:
Ah right. GitHub UI changes confused me. Regular comments just appear but the review discussion does not auto reload. Never mind.

drybjed committed:
[debops.icinga_web] Use octal number as attribute

ypid reviewed (approved):
LGTM, thanks @drybjed for your great work and working out the comments I gave!
